### PR TITLE
WIP: Accelerating KNN

### PR DESCRIPTION
**Pull request by Stolinger**
Saving knn algo and dataset hasn't been made yet.

I still see issues with the current code :

- The user can have himself as a neighbor
- The knn algo isn't completely updated when adding current user ratings line (mean_score, nb_ratings, ratings, sum_ratings)

### CODE REVIEW

File: mangaki/mangaki/utils/knn.py [hunk 1]
```diff
@@ -54,14 +54,15 @@ def get_neighbors(self, user_ids=None):
 
     def fit(self, X, y, whole_dataset=False):
         self.ratings = defaultdict(dict)
-        self.sum_ratings = defaultdict(lambda: 0)
-        self.nb_ratings = defaultdict(lambda: 0)
-        self.M = lil_matrix((self.nb_users, self.nb_works))
+        self.sum_ratings = Counter()
+        self.nb_ratings = Counter()
+        users, works = zip(*list(X))
+        self.M = coo_matrix((y,(users,works)), shape = (self.nb_users, self.nb_works)) # Might take some time, but coo is efficient for creating matrices
```
Comment (RaitoBezarius) [inline mangaki/mangaki/utils/knn.py]: :heart:
You just solved one of my OCD nightmares.

File: mangaki/mangaki/utils/recommendations.py [hunk 2]
```diff
@@ -49,27 +51,18 @@ def get_reco_algo(request, algo_name='knn', category='all'):
...
-        if request.user.is_anonymous:
-            triplets.extend([
-                (request.user.id, work_id, choice)
-                for work_id, choice in current_user_ratings(request).items()
-            ])
-
-        anonymized = dataset.make_anonymous_data(triplets)
-
-        chrono.save('make first anonymous data with {} ratings'.format(len(triplets)))
+        algo = get_algo_backup(algo_name)
```
Comment (jilljenn) [inline mangaki/mangaki/utils/recommendations.py]: À mettre dans un `try ... except FileNotFoundError:`, comme plus bas dans le fichier ;)

File: mangaki/mangaki/utils/knn.py [hunk 3]
```diff
@@ -54,14 +54,15 @@ def get_neighbors(self, user_ids=None):
...
+        self.M = self.M.tocsr() # knn.M should be CSR for faster arithmetic operations
```
Comment (jilljenn) [inline mangaki/mangaki/utils/knn.py]: I updated it because it was incorrect ^^'

File: mangaki/mangaki/utils/recommendations.py [hunk 4]
```diff
@@ -49,27 +51,24 @@ def get_reco_algo(request, algo_name='knn', category='all'):
...
-        anonymized = dataset.make_anonymous_data(triplets)
+        try:
+            algo = get_algo_backup(algo_name)
+            dataset = get_dataset_backup(algo_name)
+        except FileNotFoundError:
+            triplets = list(
+                Rating.objects.values_list('user_id', 'work_id', 'choice'))
+            chrono.save('get all %d interesting ratings' % len(triplets))
+            dataset, algo = fit_algo(algo_name, triplets)
+        already_rated_works['rating'] = already_rated_works['choice'].map(rating_values)
```
Comment (jilljenn) [inline mangaki/mangaki/utils/recommendations.py]: You cannot do this. `already_rated_works` is a list. Print it before:

    print(already_rated_works)
    print(current_user_ratings(request))

And you will figure out ;)

Comment (jilljenn) [inline mangaki/mangaki/utils/recommendations.py]: Also, `knn` is not the name of the model. It is `algo`. But you can rename it above if you want.

I'm not sure if you can `vstack` a sparse matrix with another sparse matrix. We will see.

File: mangaki/mangaki/utils/recommendations.py [hunk 5]
```diff
@@ -49,27 +51,25 @@ def get_reco_algo(request, algo_name='knn', category='all'):
...
-        anonymized = dataset.make_anonymous_data(triplets)
+        try:
+            algo = get_algo_backup(algo_name)
+            dataset = get_dataset_backup(algo_name)
+        except FileNotFoundError:
+            triplets = list(
+                Rating.objects.values_list('user_id', 'work_id', 'choice'))
+            chrono.save('get all %d interesting ratings' % len(triplets))
+            dataset, algo = fit_algo(algo_name, triplets)
+        framed_rated_works = pd.DataFrame(list(current_user_ratings(request).items()), columns = ['work_id', 'choice'])
```
Comment (jilljenn) [inline mangaki/mangaki/utils/recommendations.py]: Okay for pandas but please do not put spaces around `=` when filling arguments of a function :)

File: mangaki/mangaki/utils/recommendations.py [hunk 6]
```diff
@@ -49,27 +51,25 @@ def get_reco_algo(request, algo_name='knn', category='all'):
...
+        framed_rated_works['rating'] = framed_rated_works['choice'].map(rating_values)
+        ratings_from_user = coo_matrix((framed_rated_works['rating'],(np.zeros(len(framed_rated_works['work_id'])), framed_rated_works['work_id'])), shape = (1, algo.nb_works))
```
Comment (jilljenn) [inline mangaki/mangaki/utils/recommendations.py]: nb_rated_works = len(framed_rated_works)
    ... [0.] * nb_rated_works

is more readable I think.

(Also, directly `shape=(1, algo.nb_works)`, without spaces around.)

**Comment by jilljenn** [comment]
Premier run :

    Chrono: get rated works [4q, 10ms]
    Chrono: get all 332934 interesting ratings [5q, 1170ms]
    Chrono: loading knn and expanding with current user ratings [6q, 11710ms]
    Chrono: get neighbors [6q, 27ms]
    Chrono: fit knn-20 [7q, 292ms]
    Chrono: remove already rated [7q, 1ms]
    Chrono: compute every prediction [7q, 4ms]
    Chrono: get bulk [8q, 8ms]

Second run :

    KeyError: 8626

Hmmm peut-être qu'il y a encore des objets Work qui ne sont pas décodés comme il faut.

File: mangaki/mangaki/utils/recommendations.py [hunk 7]
```diff
@@ -34,9 +40,9 @@ def get_pos_of_best_works_for_user_via_algo(algo, dataset, user_id, work_ids, li
         pos_of_best = pos_of_best[:limit]  # Up to some limit
     return pos_of_best
 
-
 def get_reco_algo(request, algo_name='knn', category='all'):
     chrono = Chrono(is_enabled=CHRONO_ENABLED)
+    dataset = Dataset()
```
Comment (jilljenn) [inline mangaki/mangaki/utils/recommendations.py]: Is this line useless?

**Comment by jilljenn** [comment]
Fixes #396.

**Comment by jilljenn** [comment]
The above benchmark has been tried on the user that has the most ratings (3066) on the database.

File: mangaki/mangaki/utils/recommendations.py (hunk 7 above)
Comment (Stolinger) [inline mangaki/mangaki/utils/recommendations.py]: Yes, it's useless now. I'll remove it.

**Pull request merged**